File: current_session/c++/1567.cpp
#include <iostream>
#include <vector>

using namespace std;

// clang++ -std=c++17 c++/1567.cpp -o 1567 && ./1567 -v

// for each n in nums, if n == 0: start over again
// if not, focus on the continuous product itself
// e.g. [-,+,+,+,+,+,-,-,-,+,0,+,0,+,-,-,-,-,-,+,-]
// use DP
// update currentNegativeSequence, currentPositiveSequence
// update globalNegativeSequence, globalPositiveSequence

class Solution {
public:
  int getMaxLen(vector<int>& nums) {
    if (nums.size() == 0) return 0;
    int cns = 0, cps = 0, gns = 0, gps = 0;
    // initial state
    for (int i = 0; i<nums.size(); ++i) {
      if (nums[i] == 0) {
        cns = 0;
        cps = 0;
      } else if (nums[i] > 0) {
        cps += 1;
        cns = (cns == 0) ? 0 : cns+1;
      } else {
        int tmp = (cns == 0) ? 0 : cns+1;
        cns = max(cps+1, 1);
        cps = tmp;
      }
      // cout << cps << " . " << cns << endl;
      gns = max(gns, cns);
      gps = max(gps, cps);
    }
    return gps;
  }
};

int main() {
  vector<int> test1 = {1,-2,-3,4,1,-2,-3,4};
  vector<int> test2 = {0,1,-2,-3,-4};
  vector<int> test3 = {-1,-2,-3,0,1};
  cout << Solution().getMaxLen(test1) << endl;
  cout << Solution().getMaxLen(test2) << endl;
  cout << Solution().getMaxLen(test3) << endl;
}

File: current_session/java/1567.java
class Solution {
    public int getMaxLen(int[] nums) {
        int lnc = 0, lpc = 0, gnc = 0, gpc = 0;
        for (int i = 0; i < nums.length; ++i ) {
            if (nums[i] == 0) {
                lpc = 0;
                lnc = 0;
            } else if (nums[i] > 0) {
                lpc += 1;
                lnc = (lnc == 0) ? 0 : lnc+1;
            } else {
                int tmp = lpc;
                lpc = (lnc == 0) ? 0: lnc+1;
                lnc = Math.max(tmp+1, 1);
            }
            gpc = Math.max(lpc, gpc);
            gnc = Math.max(lnc, gnc);
        }
        return gpc;
    }
    // cp 1567.java Solution.java && javac Solution.java && java Solution
    public static void main(String[] args) {
        Solution sol = new Solution();
        int[] test1 = {1,-2,-3,4};
        int[] test2 = {0,1,-2,-3,-4};
        int[] test3 = {0,0,0,0,1,1,0,0,1,0,1,0,1,1,0,1,0};
        
        int ans1 = sol.getMaxLen(test1);
        int ans2 = sol.getMaxLen(test2);
        int ans3 = sol.getMaxLen(test3);
        
        System.out.println(ans1 + " should be 4");
        System.out.println(ans2 + " should be 3");
        System.out.println(ans3 + " should be 2");
    }
}